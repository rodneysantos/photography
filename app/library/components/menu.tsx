import { useState } from "react";

interface MenuItem {
	// The id of the item.
	id: number;

	// The name of the item.
	name: Item;
}

const activeMenuClasses = "bg-neutral-900 text-neutral-100";
const menuItems: MenuItem[] = [
	{ id: 0, name: "black and white" },
	{ id: 1, name: "color" },
	{ id: 2, name: "low-key" },
];

export type Item = "black and white" | "color" | "low-key";

export interface MenuProps {
	// The class name to apply to the div element.
	className?: string;

	// The ids of the active items.
	defaultActiveItems: Item[];

	// The callback to call when the active items change.
	onActiveItemsChange: (activeItems: Item[]) => void;
}

export const Menu = ({
	defaultActiveItems,
	onActiveItemsChange,
	className,
}: MenuProps) => {
	const [activeItems, setActiveItems] = useState(defaultActiveItems);

	// Toggle the active state of an item.
	// If the item is active, remove it from the active items.
	// If the item is not active, add it to the active items.
	const handleToggleItem = (name: Item) => {
		const updatedActiveItems = activeItems.includes(name)
			? activeItems.filter((item) => item !== name)
			: [...activeItems, name];

		setActiveItems(updatedActiveItems);
		onActiveItemsChange(updatedActiveItems);
	};

	return (
		<div data-testid="menu" className={className}>
			<div className="flex flex-row justify-center">
				{menuItems.map((item) => {
					const isActive = activeItems.includes(item.name);
					const className = isActive ? activeMenuClasses : "";

					return (
						<div
							key={item.id}
							data-testid={item.name}
							className={`px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900 ${className}`}
							onClick={() => handleToggleItem(item.name)}
						>
							{item.name}
						</div>
					);
				})}
			</div>
		</div>
	);
};
