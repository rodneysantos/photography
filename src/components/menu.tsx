import { createSignal } from "solid-js";
import { JSX } from "solid-js/jsx-runtime";

interface MenuItem {
	// The id of the item.
	id: number;

	// The name of the item.
	name: Item;
}

const menuItems: MenuItem[] = [
	{ id: 0, name: "black and white" },
	{ id: 1, name: "color" },
	{ id: 2, name: "low-key" },
];

export type Item = "black and white" | "color" | "low-key";

export interface MenuProps extends JSX.HTMLAttributes<HTMLDivElement> {
	// The ids of the active items.
	defaultActiveItems: Item[];

	// The callback to call when the active items change.
	onActiveItemsChange: (activeItems: Item[]) => void;

	// Whether or not to show the menu.
	show: boolean;
}

export const Menu = (props: MenuProps) => {
	const [activeItems, setActiveItems] = createSignal(props.defaultActiveItems);

	// Toggle the active state of an item.
	// If the item is active, remove it from the active items.
	// If the item is not active, add it to the active items.
	function handleToggleItem(name: Item) {
		const updatedActiveItems = activeItems().includes(name)
			? activeItems().filter((item) => item !== name)
			: [...activeItems(), name];

		setActiveItems(updatedActiveItems);
		props.onActiveItemsChange(updatedActiveItems);
	}

	return (
		<div
			data-testid="menu"
			class="transition-opacity will-change-auto"
			classList={{ "opacity-0": !props.show }}
		>
			<div class="flex flex-row justify-center">
				{menuItems.map((item) => {
					const isActive = activeItems().includes(item.name);

					return (
						<div
							data-testid={item.name}
							onClick={() => handleToggleItem(item.name)}
							class="px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900"
							classList={{ "bg-neutral-900 text-neutral-100": isActive }}
						>
							{item.name}
						</div>
					);
				})}
			</div>
		</div>
	);
};
