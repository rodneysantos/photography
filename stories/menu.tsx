import { Menu } from "../app/library/components/menu";

export const MenuSB = () => {
	return <Menu />;
};
