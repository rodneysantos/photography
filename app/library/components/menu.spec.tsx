import { beforeEach, describe, expect, it, vi } from "vitest";
import { act, render } from "@testing-library/react";
import renderer from "react-test-renderer";
import { Menu } from "./menu";

describe("menu", () => {
	const onActiveItemsChange = vi.fn();

	beforeEach(() => {
		onActiveItemsChange.mockReset();
	});

	it("renders correctly", () => {
		// arrange
		const tree = renderer
			.create(
				<Menu
					defaultActiveItems={[]}
					onActiveItemsChange={onActiveItemsChange}
				/>,
			)
			.toJSON();

		// assert
		expect(tree).toMatchSnapshot();
	});

	it("highlights the items on render", () => {
		// arrange
		const { getByTestId } = render(
			<Menu
				defaultActiveItems={["black and white", "low-key"]}
				onActiveItemsChange={onActiveItemsChange}
			/>,
		);

		// act
		const bnwItem = getByTestId("black and white");
		const lowkeyItem = getByTestId("low-key");

		// assert
		expect(bnwItem.className).toContain("bg-neutral-900");
		expect(lowkeyItem.className).toContain("bg-neutral-900");
	});

	it("higlights an item on click", () => {
		// arrange
		const { getByTestId } = render(
			<Menu
				defaultActiveItems={[]}
				onActiveItemsChange={onActiveItemsChange}
			/>,
		);

		// act
		const colorItem = getByTestId("color");
		act(() => {
			colorItem.click();
		});

		// assert
		expect(colorItem.className).toContain("bg-neutral-900");
	});

	it("unhiglights an item on click", () => {
		// arrange
		const { getByTestId } = render(
			<Menu
				defaultActiveItems={["black and white"]}
				onActiveItemsChange={onActiveItemsChange}
			/>,
		);
		const bnwItem = getByTestId("black and white");

		// act
		act(() => {
			bnwItem.click();
		});

		// assert
		expect(bnwItem.className).not.toContain("bg-neutral-900");
	});

	it("emits the active items on click", () => {
		// arrange
		const { getByTestId } = render(
			<Menu
				defaultActiveItems={[]}
				onActiveItemsChange={onActiveItemsChange}
			/>,
		);
		const colorItem = getByTestId("color");
		const lowkeyItem = getByTestId("low-key");

		// act
		act(() => {
			colorItem.click();
			lowkeyItem.click();
		});

		// assert
		expect(onActiveItemsChange).toHaveBeenCalledTimes(2);
		expect(onActiveItemsChange).toHaveBeenCalledWith(["color"]);
		expect(onActiveItemsChange).toHaveBeenCalledWith(["low-key"]);
	});
});
