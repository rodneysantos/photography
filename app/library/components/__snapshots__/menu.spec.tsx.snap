// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`menu > renders correctly 1`] = `
<div
  data-testid="menu"
>
  <div
    className="flex flex-row justify-center"
  >
    <div
      className="px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900 "
      data-testid="black and white"
      onClick={[Function]}
    >
      black and white
    </div>
    <div
      className="px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900 "
      data-testid="color"
      onClick={[Function]}
    >
      color
    </div>
    <div
      className="px-2 py-1 mx-2 text-center border-2 cursor-pointer border-neutral-900 "
      data-testid="low-key"
      onClick={[Function]}
    >
      low-key
    </div>
  </div>
</div>
`;
